src/data-model-providers/codegen/ServerClusterInterfaceRegistry.cpp

```diff
@@ -158,5 +178,59 @@ ServerClusterInterface * ServerClusterInterfaceRegistry::Get(const ConcreteClust
     return nullptr;
 }
 
+CHIP_ERROR ServerClusterInterfaceRegistry::SetContext(const ServerClusterContext & context)
+{
+    if (mContextIsValid)
+    {
+        // if there is no difference, do not re-initialize.
+        VerifyOrReturnError(mContext != context, CHIP_NO_ERROR);
+        ClearContext();
+    }
+
+    mContext         = context;
+    mContextIsValid  = true;
+    bool had_failure = false;
+
+    for (ServerClusterRegistration * registration = mRegistrations; registration != nullptr; registration = registration->next)
+    {
+        CHIP_ERROR err = registration->serverClusterInterface->Startup(&mContext);
```

Comment:
OK, so on context change this will call Startup() on clusters more than once?  That was _extremely_ non-obvious as being OK.  Why is it ok?

Reply:
if context was valid before, we call shutdown and startup again (clearcontext above will shut things down). Reasoning is: if context changes the server clusters should be notified ... so shutdown + startup seems correct.

Reply:
Provider has Startup and also Shutdown. So we need clusters to do the same, or generally have some way to shut down.

The way I approached it here is to keep the registrations after shutdown, but clear context instead (and shutdown clusters).

What I could also do is to have:
   - Register require a context (i.e Startup)
   - Shutdown to automatically unregister everything
   
It seemed non obvious that Shutdown would do a bulk unregister though as we do not unregister AAI/CHI on shutdown (but they are fairly independent singletons as is ...). Do you have a strong preference or other approach suggestions?

Reply:
Ah, I missed that we are doing a Shutdown before the second Startup.  OK.  That is certainly fine.

The fact that we do not unregister AAI/CHI on shutdown is arguably a bug, by the way, so I am not sure about the right behavior there.  But I don't have strong preferences, no, and don't think we should block here on figuring it out.  Maybe file a followup issue on sorting out the lifetime story here?  

For example: why would we ever "change context" on a registry at all?  Seems to me that instead of having SetContext/ClearContext we should just have Init() (taking context) and Shutdown on the registry, and the codegen provider calls those from its own init/shutdown.  But that can be a followup.